OpenDreamRuntime/Procs/DebugAdapter/DreamDebugManager.cs

```diff
@@ -66,75 +97,109 @@ private struct ActiveFunctionBreakpoint {
+        foreach (var bp in fileSlots[line].Breakpoints ?? Enumerable.Empty<ActiveBreakpoint>()) {
+            if (TestBreakpoint(bp)) {
+                hit.Add(bp.Id);
             }
         }
         if (hit.Any()) {
-            Stop(new StoppedEvent {
+            Output($"Breakpoint hit at {state.CurrentSource}:{line}");
+            Stop(state.Thread, new StoppedEvent {
                 Reason = StoppedEvent.ReasonBreakpoint,
-                ThreadId = state.Thread.Id,
-                AllThreadsStopped = true,
                 HitBreakpointIds = hit,
             });
         }
     }
 
     public void HandleProcStart(DMProcState state) {
         var hit = new List<int>();
-        if (functionBreakpoints != null) {
-            foreach (var bp in functionBreakpoints[(state.Proc.OwningType.PathString, state.Proc.Name)]) {
-                if (bp.Condition is null && bp.HitCondition is null) {
+        if (_possibleFunctionBreakpoints?.GetValueOrDefault((state.Proc.OwningType.PathString, state.Proc.Name)) is FunctionBreakpointSlot slot) {
+            foreach (var bp in slot.Breakpoints) {
+                if (TestBreakpoint(bp)) {
                     hit.Add(bp.Id);
-                    break;
                 }
             }
         }
         if (hit.Any()) {
-            Logger.Info($"Function breakpoint hit at {state.Proc.OwningType.PathString}::{state.Proc.Name}");
-            Stop(new StoppedEvent {
+            Output($"Function breakpoint hit at {state.Proc.OwningType.PathString}::{state.Proc.Name}");
+            Stop(state.Thread, new StoppedEvent {
                 Reason = StoppedEvent.ReasonFunctionBreakpoint,
-                ThreadId = state.Thread.Id,
-                AllThreadsStopped = true,
                 HitBreakpointIds = hit,
             });
         }
     }
 
     public void HandleException(DreamThread thread, Exception exception) {
-        _exception = exception;
-        Stop(new StoppedEvent {
-            Reason = StoppedEvent.ReasonException,
-            ThreadId = thread.Id,
-            AllThreadsStopped = true,
-        });
+        if (breakOnRuntimes) {
+            _exception = exception;
+            Output("Stopped on exception");
+            Stop(thread, new StoppedEvent {
+                Reason = StoppedEvent.ReasonException,
+            });
+        }
     }
 
-    private void Stop(StoppedEvent stoppedEvent) {
-        if (_adapter == null || !_adapter.AnyClientsConnected())
+    private bool TestBreakpoint(ActiveBreakpoint bp) => bp.Condition is null && bp.HitCondition is null;
+
+    // Utilities
+    private void Output(string message, string category = OutputEvent.CategoryConsole) {
+        _adapter?.SendAll(new OutputEvent(category, $"{message}\n"));
+    }
+
+    private bool CanStop() => _adapter != null && _adapter.AnyClientsConnected() && !_terminated;
+
+    private void Stop(DreamThread? thread, StoppedEvent stoppedEvent) {
+        if (_adapter == null || !CanStop())
```

Comment:
`!CanStop()` already checks `_adapter == null`
```suggestion
        if (!CanStop())
```

Reply:
The nullability linter can't see through this and complains about an unprotected access of `_adapter` later in the method..